File: assign2/haskell/src/FibCalculator.hs
module FibCalculator where

    iterativeFib 0 = [1]
    iterativeFib 1 = [1, 1]
    iterativeFib n = (iterate (\list  -> list ++ [(last list) + last (init list)]) [1,1] !! (n - 1))
    
    recursiveFib 0 = [1]
    recursiveFib 1 = [1, 1]
    recursiveFib n = list ++ [sum]
        where
            list = recursiveFib(n - 1)
            sum = list!!(n - 2) + list!!(n - 1)

    recursiveIterativeFib n = recurse n [1, 1]
        where
            recurse n list  | n == 0    = [1]
                            | n == 1    = list
                            | otherwise = recurse (n - 1) (list ++ [(last list) + last (init list)])

File: assign2/clojure/src/fibcalculator.clj
(ns fibcalculator)

(defn iterativeFib [n] 
  (case n
    0 [1]
    1 [1, 1]
    (nth (iterate (fn [flist] (conj flist (reduce +' (take-last 2 flist)))) '[1, 1]) (dec n))))

(def recursiveFib 
  (fn [n]
    (case n
      0 [1]
      1 [1, 1]
      ((fn [list] (conj list (reduce +' (take-last 2 list)))) (recursiveFib (dec n))))))

(defn recursiveiterativeFib [num]
  (letfn [(recursiveCall [number list]
          (case number
              0 [1]
              1 list
              (recursiveCall (dec number) (conj list (reduce +' (take-last 2 list))))
          ))]
  (recursiveCall num [1, 1]))
)